### PR TITLE
rlat and rlon reprojection

### PR DESCRIPTION
**Pull request by adanb13**
### Changes

- User provides `rlon` and `rlat` in WGS84, provider re-projects using native CRS

### Example Links
_NOTE: Need to use slight range for `rlat/rlon` as `nearest` not yet implemented_

**Toronto**
http://infopool-dev1.cmc.ec.gc.ca:5089/collections/weong:hrdps:tt/coverage?subset=rlon(-79.421371:-72),rlat(43.780918:45),level(1:1)&datetime=2023-03-09T00:00:00.000000000&f=json

**Ottawa**
http://infopool-dev1.cmc.ec.gc.ca:5089/collections/weong:hrdps:tt/coverage?subset=rlon(-75.695000:-74),rlat(43.780918:45),level(1:1)&datetime=2023-03-09T00:00:00.000000000&f=json

**Edmonton**
http://infopool-dev1.cmc.ec.gc.ca:5089/collections/weong:hrdps:tt/coverage?subset=rlon(-113.4937:-112),rlat(53.5461:54),level(1:1)&datetime=2023-03-09T00:00:00.000000000&f=json

**New Brunswick**
http://infopool-dev1.cmc.ec.gc.ca:5089/collections/weong:hrdps:tt/coverage?subset=rlon(-66.159668:-65),rlat(46.498390:47),level(1:1)&datetime=2023-03-09T00:00:00.000000000&f=json

**Prince Edward Island**
http://infopool-dev1.cmc.ec.gc.ca:5089/collections/weong:hrdps:tt/coverage?subset=rlon(-63.000000:-62),rlat(46.250000:47),level(1:1)&datetime=2023-03-09T00:00:00.000000000&f=json

**Yukon**
http://infopool-dev1.cmc.ec.gc.ca:5089/collections/weong:hrdps:tt/coverage?subset=rlon(-131.352539:-130),rlat(62.350357:63),level(1:1)&datetime=2023-03-09T00:00:00.000000000&f=json

### TO DO

- Have Shaheed confirm provider now returns data as required
- implement `nearest` method for point queries
- _(Improvement)_ implement reprojection to `bbox` queries allowing us to use `.sel` instead of `.where` which is much faster and gets rid for the need to use `eval`
- Change CoverageJSON `axes` information to WGS84 user given input instead of current reprojected coordinates being given

### CODE REVIEW

**Comment by adanb13** [comment]
with reference to "time query fix": (just pushed changes that would make it so nearest also activates on an invalid datetime value, to undo and only have it activate for rlon/rlon values I would revert to the previous commit)

File: msc_pygeoapi/provider/hrdps_zarr.py [hunk 1]
```diff
@@ -38,6 +38,7 @@
     ProviderInvalidQueryError,
     ProviderNoDataError
 )
+from pyproj import CRS, Transformer
```
Comment (tomkralidis) [inline msc_pygeoapi/provider/hrdps_zarr.py:41]: move to after the `dask.array` import.

File: msc_pygeoapi/provider/hrdps_zarr.py [hunk 2]
```diff
@@ -448,6 +493,45 @@ def _gen_domain_axis(self, data):
     return aa, all_dims
 
 
+def _make_where_str(first_query: dict, new_query: dict):
+    """
```
Comment (tomkralidis) [inline msc_pygeoapi/provider/hrdps_zarr.py:497]: Space out all docstrings, i.e.:

```python
    """
    Helper function to make a where query string for
    nearest method when exact coordinates are not found

    :param first_query: original_dict of query
    :param new_query: transformed_dict of query

    :returns: where query string for use in .where()
    """
```

File: msc_pygeoapi/provider/hrdps_zarr.py [hunk 3]
```diff
@@ -448,6 +493,45 @@ def _gen_domain_axis(self, data):
...
+    where_str = ''
+    for key in first_query.keys():
+        if key not in new_query.keys():
+            where_str += f'(self._data.{key}>={first_query[key].start}) & '
+            where_str += f'(self._data.{key}<={first_query[key].stop}) & '
+    LOGGER.debug(f'Where query string: {where_str}')
+    return where_str[:-3]
+
+
+def _convert_subset_to_crs(new_lat: slice, new_lon: slice, crs):
```
Comment (tomkralidis) [inline msc_pygeoapi/provider/hrdps_zarr.py:514]: Space out docstring.

File: msc_pygeoapi/provider/hrdps_zarr.py [hunk 4]
```diff
@@ -448,6 +493,45 @@ def _gen_domain_axis(self, data):
...
+def _make_where_str(first_query: dict, new_query: dict):
+    """
+    Helper function to make a where query string for
+    nearest method when exact coordinates are not found
+    :param first_query: original_dict of query
+    :param new_query: transformed_dict of query
+    :returns: where query string for use in .where()
+    """
+
+    where_str = ''
```
Comment (tomkralidis) [inline msc_pygeoapi/provider/hrdps_zarr.py:505]: I think we are trying to build up a query string here?  Possible alternative:

```python
    query_params = []

    for key in first_query.keys():
        if key not in new_query.keys():
            query_params.append(f'(self._data.{key}>={first_query[key].start})')
            query_params.append(f'(self._data.{key}<={first_query[key].stop})')

    query_string = ' & '.join(query_params)
    LOGGER.debug(f'Where query string: {query_string}')
    return query_string
```

File: msc_pygeoapi/provider/hrdps_zarr.py [hunk 5]
```diff
@@ -448,6 +493,45 @@ def _gen_domain_axis(self, data):
     return aa, all_dims
 
 
+def _make_where_str(first_query: dict, new_query: dict):
```
Comment (tomkralidis) [inline msc_pygeoapi/provider/hrdps_zarr.py:496]: Provide a return type in the function annotation, so:


```python
def _make_where_str(first_query: dict, new_query: dict) -> str:
```

File: msc_pygeoapi/provider/hrdps_zarr.py (hunk 3 above)
Comment (tomkralidis) [inline msc_pygeoapi/provider/hrdps_zarr.py:514]: Provide a return type in the function annotation, so:

```python
def _make_where_str(first_query: dict, new_query: dict) -> Tuple:
```

In this case you will have to add the following import:

```python
from typing import Tuple
```

**Review by tomkralidis — CHANGES_REQUESTED** [review]

**Pull request merged**